packages/debug/src/browser/debug-session.tsx

```diff
@@ -368,14 +368,19 @@ export class DebugSession implements CompositeTreeElement {
     }
 
     protected async runInTerminal({ arguments: { title, cwd, args, env } }: DebugProtocol.RunInTerminalRequest): Promise<DebugProtocol.RunInTerminalResponse['body']> {
-        return this.doRunInTerminal({ title, cwd, shellPath: args[0], shellArgs: args.slice(1), env });
+        const terminal = await this.doCreateTerminal({ title, cwd, env });
+        terminal.sendText(args.join(' ') + '\n');
```

Comment:
@akosyakov This part is causing us issues, but it is also a bad thing to do to send un-escaped commands to a shell like this. IIUC, the idea is that we want to reuse a shell widget to display the output. But in this case it seems like we try to reuse actual shell processes, instead of reusing widgets.

Since this is broken for us, I will start trying to look at what is required to reuse shell widgets, do you have any opinions and/or pointers for me?

cc @marcdumais-work

Reply:
Maybe better to open an issue, will do that.

Reply:
here is VS Code implementation:https://github.com/microsoft/vscode/blob/43e3581aa635406ada2d8138d7ee242ddabefc1c/src/vs/workbench/contrib/debug/node/terminals.ts#L79

We can reuse it.

Reply:
I don't like this, but it may be the easiest solution here. But why would we convert an nice array of arguments into one big escaped string though? We can spawn processes directly from a command + arguments. Do debug adapters expect that their clients will provide variable-expansion or other shell-specific behaviors?

Reply:
After looking more into it, we do want to use shells. Things like running `node` without specifying a full path wouldn't work otherwise. Not sure we want to try to always reuse the same shell process though. In VS Code they check before reusing a shell that it doesn't have child processes running, this seems sketchy. I would rather make sure the shell dies when the command is executed, and reuse the widget if process is dead. WDYT?

Reply:
Don't get it. Here is a solution, which we can copy and fix the issue, and then focus on something else instead of spending time reinventing our solution. Plus it makes it easier to catch up later if there are some bug fixes.

> Not sure we want to try to always reuse the same shell process though. In VS Code they check before reusing a shell that it doesn't have child processes running, this seems sketchy. I would rather make sure the shell dies when the command is executed, and reuse the widget if process is dead. WDYT?

Sorry, I don't know how it is related to escaping. We start a new terminal each time for now. Please file a new issue if there is a mismatch from user perspective with steps how to reproduce in Theia and VS Code.

Reply:
Terminals are reused when handling `runInTerminal` requests:

![theia-vscode-python-ext](https://user-images.githubusercontent.com/14182238/70247029-aadd4980-1746-11ea-9eef-5c06c3805e65.gif)

Reply:
But apparently VS Code has the same issue when running the same command, user can pollute the shell that the extension tries to send commands to... Ok.